apps/credential-service/lib/server.ts

```diff
@@ -0,0 +1,160 @@
+
+/** Generate a temporary access token with access to the requested repository. */
+export async function generateAccessToken(
+  owner: string,
+  repo: string,
+): Promise<AccessTokenAndRevocation> {
+  /** The github client used for generating the token. */
+  const github = new Octokit({
+    authStrategy: createAppAuth,
+    auth: {appId: GITHUB_APP_ID, privateKey: GITHUB_APP_PEM},
+  });
+  /** The specific installation id for the provided repository. */
+  const {id: installation_id} = (await github.apps.getRepoInstallation({owner, repo})).data;
+  /** A temporary github access token. */
+  const {token} = (await github.rest.apps.createInstallationAccessToken({installation_id})).data;
+
+  return {
+    token,
+    revokeToken: async () => await github.apps.revokeInstallationAccessToken(),
+  };
+}
+
+/**
+ * WebSocket handler to generate temporary Github access token.
+ *
+ * The access token is automatically revoked when the websocket is closed.
+ */
+async function wsHandler(ws: WebSocket, req: IncomingMessage) {
+  /** Whether the websocket heartbeat check is still alive. */
+  let receivedHeartbeatResponse: boolean;
+  /** The interval instance for checking the heartbeat. */
+  let heartbeatInterval = setInterval(checkHeartbeat, heartBeatIntervalLength);
+  /**
+   * The repository name and owner for the request.
+   * Note: We safely case the header type as having these fields because they are checked prior
+   *       to the WebSocket handler function being invoked.
+   */
+  const {NgDevTargetRepoOwner: owner, NgDevTargetRepoName: repo} = req.headers as {
+    NgDevTargetRepoName: string;
+    NgDevTargetRepoOwner: string;
+  };
+
+  /** The temporary Github access token and function to revoke the token.. */
+  const {token, revokeToken} = await generateAccessToken(owner, repo);
+  /** Check to make sure the heartbeat is still alive. */
+  function checkHeartbeat() {
+    if (receivedHeartbeatResponse === false) {
+      ws.close(1008, 'Cannot find socket via heartbeat check');
+      ws.terminate();
+    }
+    receivedHeartbeatResponse = false;
+    ws.ping();
+  }
+
+  /**
+   * Cleans up the state of the function when the WebSocket is completed.
+   */
+  async function complete() {
+    await revokeToken();
+    clearInterval(heartbeatInterval);
+  }
+
+  // Ensure that cleanup is done when the WebSocket closes.
+  ws.on('close', complete);
+  ws.on('error', complete);
+
+  // Handle the pong response from the websocket client, updating the heartbeat as alive.
+  ws.on('pong', () => (receivedHeartbeatResponse = true));
+
+  // Send the temporary Github token to the websocket client
+  ws.send(token);
+}
+
+/**
+ * Handle upgrade requests before the websocket is used. Enforces authentication mechanisms
+ * and ensuring the required data is present in the request.
+ */
+async function upgradeHandler(req: IncomingMessage, socket: Duplex, head: Buffer) {
+  try {
+    if (!authorizationRegex.test(req.headers.authorization!)) {
```

Comment:
Super nit (while I leave actual comments): Is the exclamation needed? generally a good thing to avoid

Reply:
Switched it to a `|| ''`, which ends up functioning the same for `Regex.test`